### ISSUE TITLE
Suppressions DTD not found

### ISSUE THREAD

**Issue by ddidier** [checkstyle/checkstyle#10788]
Hi,

I'm using CheckStyle with Gradle and I'm using `SuppressionsFilter`.

- without `SuppressionsFilter` and `./gradlew build` => pass
- without `SuppressionsFilter` and `./gradlew build -Dcheckstyle.enableExternalDtdLoad=true` => pass
- with `SuppressionsFilter` and `./gradlew build` => fail
- with `SuppressionsFilter` and `./gradlew build -Dcheckstyle.enableExternalDtdLoad=true` => pass

So this error is related to https://github.com/checkstyle/checkstyle/issues/6474.

The Gradle plugin seems to use the Ant task (see logs below) and it looks like the Ant task is not getting the DTD from the CheckStyle JAR but from the internet only. So this may be a CheckStyle issue rather than a Gradle one.

```
Caused by: java.lang.NullPointerException: Cannot invoke "java.util.Map.entrySet()" because "this.fTableOfNOTATIONAttributeNames" is null
        at java.xml/com.sun.org.apache.xerces.internal.impl.dtd.XMLDTDProcessor.endDTD(XMLDTDProcessor.java:1301)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.endEntity(XMLDTDScannerImpl.java:658)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLEntityManager.endEntity(XMLEntityManager.java:1507)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLEntityScanner.load(XMLEntityScanner.java:1925)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLEntityScanner.skipSpaces(XMLEntityScanner.java:1664)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.skipSeparator(XMLDTDScannerImpl.java:2088)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.scanDecls(XMLDTDScannerImpl.java:2064)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDTDScannerImpl.scanDTDExternalSubset(XMLDTDScannerImpl.java:299)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.dispatch(XMLDocumentScannerImpl.java:1165)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$DTDDriver.next(XMLDocumentScannerImpl.java:1040)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$PrologDriver.next(XMLDocumentScannerImpl.java:917)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:541)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:888)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:824)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1224)
        at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:635)
        at com.puppycrawl.tools.checkstyle.XmlLoader.parseInputSource(XmlLoader.java:86)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.getSuppressionLoader(SuppressionsLoader.java:283)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.loadSuppressions(SuppressionsLoader.java:238)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.loadSuppressions(SuppressionsLoader.java:224)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionFilter.finishLocalSetup(SuppressionFilter.java:274)
        at com.puppycrawl.tools.checkstyle.api.AutomaticBean.configure(AutomaticBean.java:197)
        at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:468)
        at com.puppycrawl.tools.checkstyle.api.AutomaticBean.configure(AutomaticBean.java:201)
        at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.createRootModule(CheckstyleAntTask.java:424)
        at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.realExecute(CheckstyleAntTask.java:320)
        at com.puppycrawl.tools.checkstyle.ant.CheckstyleAntTask.execute(CheckstyleAntTask.java:302)
        at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:99)
        ... 142 more
```

**Comment by rnveach** [comment]
Please provide the XML file and everything else needed to reproduce the issue. https://checkstyle.org/report_issue.html#How_to_report_a_bug.3F

**Comment by ddidier** [comment]
I found the culprit. I'm using XML entities in the `suppressions.xml` file, i.e.:

```xml
<!DOCTYPE suppressions PUBLIC
    "-//Puppy Crawl//DTD Suppressions 1.2//EN"
    "http://www.puppycrawl.com/dtds/suppressions_1_2.dtd" [
    <!ENTITY csiMain "src.main.java.checkstyle.issue">
    ]>
```

- No XML entity and `./gradlew build -Dcheckstyle.enableExternalDtdLoad=true` => pass
- No XML entity and `./gradlew build` => pass
- XML entity and `./gradlew build -Dcheckstyle.enableExternalDtdLoad=true` => pass
- XML entity and `./gradlew build` => fail

I attach a small Gradle project that demonstrate this behavior: [checkstyle-issue-10788.tar.gz](https://github.com/checkstyle/checkstyle/files/7129745/checkstyle-issue-10788.tar.gz)

**Comment by strkkk** [comment]
@ddidier does it mean that issue can be closed?

**Comment by ddidier** [comment]
I would say no since this behavior seems unexpected. Why does it work in one case and not in the other?

**Comment by rnveach** [comment]
> Why does it work in one case and not in the other?

What 2 cases are you referring to? If your suppression file has entities, then it makes sense that it passes without the filter and fails with it when you do not provide `checkstyle.enableExternalDtdLoad`.

Checkstyle disallows the features by default in XMLs without that extra parameter:
* load-external-dtd
* external-general-entities
* external-parameter-entities

I assume your file may be falling under the parameter entities.

**Comment by romani** [comment]
We do not support Gradle plugin, to keep issue valid you need to repoduce problem by cli, see link above a how to report issue.

**Comment by ddidier** [comment]
> Checkstyle disallows the features by default in XMLs without that extra parameter:
> 
>     * load-external-dtd
>     * external-general-entities
>     * external-parameter-entities
> 
> I assume your file may be falling under the parameter entities.

That makes sense. I wasn't aware of the second and third limitations. I looked quickly in the documentation but couldn't find these items, and specifically not in https://checkstyle.sourceforge.io/config_system_properties.html#Enable_External_DTD_load. Do you want me to open a new issue for this? You can close this one.

Thanks

**Comment by romani** [comment]
Better to open new issue
